src/editor/hotkeys.cpp

```diff
@@ -1,13 +1,70 @@
 #include "codeeditor.h"
 
+#include <QRegularExpression>
+
+QString CodeEditor::tabs="";
+
+bool CodeEditor::insidebracket(){
+    QTextCursor cursor = this->textCursor();
+    if(cursor.position()!=this->document()->toPlainText().size()){
+        cursor.movePosition(QTextCursor::PreviousCharacter);
+        int currentpos = cursor.position();
+        QString prev =this->document()->toPlainText().at(currentpos);
+        cursor.movePosition(QTextCursor::NextCharacter);
+        currentpos = cursor.position();
+        QString next = this->document()->toPlainText().at(currentpos);
+        if(prev=="{" && next=="}")return true;
+    }
+    return false;
+}
+
+
+
+void CodeEditor::autotab(){
+    QString text(this->document()->toRawText());
+    int lbrackets=0 ;
+    int rbrackets=0 ;
+    QTextCursor crs=this->textCursor();
+    for (int c=0; c<crs.position();c++){
+        if(text.at(c)=="{") lbrackets++;
+        if(text.at(c)=="}") rbrackets++;
+    }
+    tabs="";
+    int difference=lbrackets-rbrackets;
+
+    for(int i=0; i<difference;i++)
+    {
+      tabs.append("\t");
+    }
+
+}
+
 void CodeEditor::keyPressEvent(QKeyEvent *e)
 {
+   static bool pressSlesh=false;
```

Comment:
This function too complicated.  Please avoid code duplication. Split code between functions. It isn't easy to maintain and read so complicated functions